### PR TITLE
ID3v2: Fix about number pair of track and disk

### LINKED ISSUE #145: ID3v2 track & track_total frames are written incorrectly (disk and disk_total probably too)
### Reproducer

I tried this code:

```rust
	let mut tagged_file = Probe::open(&opt.path)
		.expect("ERROR: Bad path provided!")
		.read()
		.expect("ERROR: Failed to read file!");

	let tag = tagged_file.primary_tag_mut().unwrap();
        tag.set_track(9);
        tag.set_track_total(13);
        tag.save_to_path(&opt.path).unwrap()
```

### Summary

Track numbers in id3v2 are stored in the TRCK frame, encoded like this `{track_number}/{track_total}`. Lofty-rs tag currently writes track and track_total as two separate TRCK frames. It looks like this in a hex editor: `TRCK.......9TRCK.......13`

### Suggested fix:
in `src/id3/v2/tag.rs` on line 763 in tag_frames add special handling for these frames.

### Expected behavior

Lofty writing the TRCK frame correctly. Like this: `TRCK.......9/13`

### Assets

ID3v2 Spec (ID3 website is currently broken):
https://web.archive.org/web/20191202011405/http://id3.org/id3v2.4.0-frames

### PR DESCRIPTION
**Pull request by mikanbako**
This PR fixes #145.

TODOs

- [x] Rebase to the main branch for conflict.
- [x] Fix id3::v2::tag::tag_frames() that called from tag::Tag::save_to().
- [x] Add tests about frame values.
- [x] Improve parsing number pair.

### CODE REVIEW

File: src/id3/v2/tag.rs [hunk 1]
```diff
@@ -285,6 +288,30 @@ impl ID3v2Tag {
 
 		(None, None)
 	}
+
+	fn insert_frame(&mut self, item: TagItem) {
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:292]: The purpose of this function is to insert an _item_ by mapping it to the underlying representation, but it does not (always) insert a frame -> `insert_item()`

File: src/id3/v2/tag.rs [hunk 2]
```diff
@@ -285,6 +288,30 @@ impl ID3v2Tag {
...
+		fn set_number<F: FnMut(u32)>(item: TagItem, mut setter: F) {
+			let number = item.into_value().text().map(str::parse::<u32>);
+
+			if let Some(Ok(number)) = number {
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:296]: Log a warning instead of silently ignoring values that couldn't be parsed successfully?

File: src/id3/v2/tag.rs [hunk 3]
```diff
@@ -349,6 +376,14 @@ fn new_picture_frame(picture: Picture, flags: FrameFlags) -> Frame<'static> {
 	}
 }
 
+fn merge_num_pair<N, T>(number: N, total: T) -> String
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:379]: - `format_number_pair()`? No need for abbreviations here.
- Using optional arguments would eliminate redundant code in setters and make this function more versatile **and** testable in isolation.

File: src/id3/v2/tag.rs [hunk 4]
```diff
@@ -872,10 +915,56 @@ impl<'a> Id3v2TagRef<'a, std::iter::Empty<FrameRef<'a>>> {
...
+		id: &'a str,
+	) -> Option<FrameRef<'a>> {
+		let mut value: Option<Cow<'_, str>> = None;
+
+		if let Some(number) = number {
+			value = Some(Cow::Borrowed(number));
+		}
+
+		if let Some(total) = total {
+			let number = value.unwrap_or_else(|| Cow::Owned(DEFAULT_NUMBER.to_string()));
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:930]: A `const &str` representation for `DEFAULT_NUMBER` would avoid an unecessary allocation -> `value.as_deref().unwrap_or(DEFAULT_NUMBER_STR)`.

With the proposed changes and extension of `merge_num_pair` this might even become obsolete.

File: src/id3/v2/tag.rs [hunk 5]
```diff
@@ -872,10 +915,56 @@ impl<'a> Id3v2TagRef<'a, std::iter::Empty<FrameRef<'a>>> {
...
+
+			FrameRef {
+				id: frame.id,
+				value: Cow::Owned(frame.value),
+				flags: frame.flags,
+			}
+		})
+	}
+
+	let number_pair_keys = &[
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:946]: This should be extracted as a constant of the module. Probably near the other _number_ constants. Serves as additional documentation.

File: src/id3/v2/tag.rs [hunk 6]
```diff
@@ -1634,4 +1723,103 @@ mod tests {
 
 		assert_eq!(id3v2, reparsed);
 	}
+
+	#[test]
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:1727]: The tests only cover a single case. Missing are test for only number and only total.

Comment (uklotzde) [reply src/id3/v2/tag.rs:1727]: I am also curious to know what happens if the frame value

- is empty (contains only whitespace)
- "/"
- "/1"
- "1/"

This should be handled by lightweight unit tests, no need for heavy-weight file tests.

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by mikanbako** [comment]
@uklotzde Thank you for review. I will fix this PR.

**Comment by uklotzde** [comment]
> @uklotzde Thank you for review. I will fix this PR.

Take my proposals with a grain of salt. All my remarks in PRs are intended as additional input from an outside viewer who has a different perspective on the topic. They should support you to validate and rethink your design decisions or to reveal blank spaces and edge cases.

A seemingly dumb question is often the most helpful feedback you could get for improving the code.

**Comment by mikanbako** [comment]
I fixed my existing code at first.

I will add tests about the frame values.

@uklotzde Thank you for comment. You have helped me to improve my PR more.

File: src/id3/v2/tag.rs [hunk 7]
```diff
@@ -25,6 +26,8 @@ const COMMENT_FRAME_ID: &str = "COMM";
 const V4_MULTI_VALUE_SEPARATOR: char = '\0';
 const NUMBER_PAIR_SEPARATOR: char = '/';
 
+const DEFAULT_NUMBER: u32 = 1;
```
[the remark this answers is not in the document]
Comment (mikanbako) [reply src/id3/v2/tag.rs]: I fixed this part.

File: src/id3/v2/tag.rs (hunk 1 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs]: I fixed this part.

File: src/id3/v2/tag.rs (hunk 2 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs]: I fixed this part.

File: src/id3/v2/tag.rs (hunk 3 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs]: I fixed this part.

File: src/id3/v2/tag.rs (hunk 4 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs]: I did not fixed it. Because `merge_num_pair` was changed.

File: src/id3/v2/tag.rs (hunk 5 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs]: I fixed this part.

File: src/id3/v2/tag.rs (hunk 6 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs:1722]: I added tests about the missing tests at first.

I will add tests about the frame values.

**Comment by mikanbako** [comment]
I fixed a log message only.

I will add tests about the frame values.

File: src/id3/v2/tag.rs [hunk 8]
```diff
@@ -285,6 +297,52 @@ impl ID3v2Tag {
...
+					}
+				}
+			},
+		};
+	}
+
+	fn insert_number_pair(&mut self, id: &'static str, number: Option<u32>, total: Option<u32>) {
+		let content = format_number_pair(number, total);
+
+		match content {
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:339]: Using `if let Some() = ` would be more idiomatic. I remember that clippy warns when matching on `Option`.

**Comment by mikanbako** [comment]
I added tests about the frame values and fixed the code to except invalid values.

File: src/id3/v2/tag.rs (hunk 8 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs]: Thank you for comment. I fixed the code.

File: src/id3/v2/tag.rs [hunk 9]
```diff
@@ -1634,4 +1734,278 @@ mod tests {
...
+			assert!(tag.disk_total().is_none());
+		}
+
+		assert_invalid("");
+		assert_invalid(" ");
+		assert_invalid("/");
+		assert_invalid("/1");
+		assert_invalid("1/");
+		assert_invalid("a/b");
+		assert_invalid("1 / 2");
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:2006]: I would expect that leading/trailing whitespaces around the numbers are trimmed.

https://en.wikipedia.org/wiki/Robustness_principle

File: src/id3/v2/tag.rs [hunk 10]
```diff
@@ -1634,4 +1734,278 @@ mod tests {
...
+		assert_invalid("1/2/3");
+		assert_invalid("1//2");
+		assert_invalid(" 1/2 ");
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:2009]: This input should also be considered as valid.

File: src/id3/v2/tag.rs [hunk 11]
```diff
@@ -547,21 +607,38 @@ impl SplitAndMergeTag for ID3v2Tag {
...
-			if let Some(total) = split.next() {
-				tag.items
-					.push(TagItem::new(total_key, ItemValue::Text(total.to_string())))
-			}
+			let number = split.next()?;
+			let total = split.next();
+
+			let is_valid = if let Some(total) = total {
+				content.len() == number.len() + 1 + total.len()
+					&& str::parse::<u32>(number).is_ok()
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:621]: By extracting a `parse_number()` function you could handle the whitespace trimming consistently. I consider "total" as a special kind of number, so the naming should be reasonable and we do not need to name it `parse_number_or_total()` .

**Comment by mikanbako** [comment]
@uklotzde Thank you for comments. I will modify my code.

Before I do that, please allow to confirm the details about valid or invalid frames.

I understood the following table, is it right?

<table>
<th>frame value</th>
<th>valid or invalid</th>
<th>number</th>
<th>total
<tr><td>/1</td><td>invalid</td><td>-</td><td>-</td></tr>
<tr><td>1/</td><td>invalid</td><td>-</td><td>-</td></tr>
<tr><td>1&nbsp;/&nbsp;2</td><td>valid</td><td>1</td><td>2</td></tr>
<tr><td>1/2/3</td><td>valid</td><td>1</td><td>2</td></tr>
<tr><td>1//2</td><td>valid</td><td>1</td><td>2</td></tr>
<tr><td>&nbsp;1/2&nbsp;</td><td>valid</td><td>1</td><td>2</td></tr>
<tr><td>a/b&nbsp;</td><td>invalid</td><td>-</td><td>-</td></tr>
<tr><td>1/b&nbsp;</td><td>invalid</td><td>-</td><td>-</td></tr>
<tr><td>a/1&nbsp;</td><td>invalid</td><td>-</td><td>-</td></tr>
</table>

**Comment by uklotzde** [comment]
> Before I do that, please allow to confirm the details about valid or invalid frames.

I suggest to only ignore whitespace, because the overarching object is to preserve all (meaningful) information. Consequently both `1/2/3` and `1//2` should **not** be parsed as a pair of numbers, because we cannot reason about the intention of the user.

I am unsure about `1/` and `/1` which could both be interpreted as a missing numbers (defaulting to 0) when using a sloppy formatter that unconditionally inserts the separator.

**Comment by mikanbako** [comment]
@uklotzde Thank you for the comment. 

I will modify my code to ignore whitespaces.

For the present, `1/` and `/1` will be treated as invalid.

In other words, I plan to modify my code for the following table.

frame value | valid or invalid | number | total
-- | -- | -- | --
/1 | invalid | - | -
1/ | invalid | - | -
1 / 2 | valid | 1 | 2
1/2/3 | invalid | - | -
1//2 | invalid | - | -
&nbsp;1/2&nbsp; | valid | 1 | 2
a/b | invalid | - | -
1/b | invalid | - | -
a/1 | invalid | - | -

**Comment by mikanbako** [comment]
I modified my code to ignore whitespaces while parsing a number pair.

File: src/id3/v2/tag.rs [hunk 12]
```diff
@@ -547,21 +607,50 @@ impl SplitAndMergeTag for ID3v2Tag {
 		fn split_pair(
 			content: &str,
 			tag: &mut Tag,
-			current_key: ItemKey,
+			number_key: ItemKey,
 			total_key: ItemKey,
 		) -> Option<()> {
+			fn parse_number(source: Option<&str>) -> Option<&str> {
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:613]: This function should accept `source: &str`. Otherwise you are implicitly re-implementing `Option::and_then`.

File: src/id3/v2/tag.rs [hunk 13]
```diff
@@ -547,21 +607,50 @@ impl SplitAndMergeTag for ID3v2Tag {
...
-				.push(TagItem::new(current_key, ItemValue::Text(current)));
 
-			if let Some(total) = split.next() {
-				tag.items
-					.push(TagItem::new(total_key, ItemValue::Text(total.to_string())))
-			}
+			let number_source = split.next()?;
+			let number = parse_number(Some(number_source))?;
+			let total_source = split.next();
+			let total = parse_number(total_source);
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:631]: `let total = total_source.and_then(parse_number)`

File: src/id3/v2/tag.rs [hunk 14]
```diff
@@ -547,21 +607,50 @@ impl SplitAndMergeTag for ID3v2Tag {
...
+			number_key: ItemKey,
 			total_key: ItemKey,
 		) -> Option<()> {
+			fn parse_number(source: Option<&str>) -> Option<&str> {
+				let number = source?.trim();
+
+				if str::parse::<u32>(number).is_ok() {
+					Some(number)
+				} else {
+					log::warn!("{number:?} could not be parsed as a number.");
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:619]: Logging a warning if the input is empty is not desired.

**Comment by mikanbako** [comment]
@uklotzde Thank you for the advice. I fixed my code.

File: src/id3/v2/tag.rs [hunk 15]
```diff
@@ -547,21 +607,54 @@ impl SplitAndMergeTag for ID3v2Tag {
 		fn split_pair(
 			content: &str,
 			tag: &mut Tag,
-			current_key: ItemKey,
+			number_key: ItemKey,
 			total_key: ItemKey,
 		) -> Option<()> {
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:612]: Using `Option<()>` for early returns is sometimes handy but also confusing. Better return nothing and use if-let-else statements for early returns.

File: src/id3/v2/tag.rs [hunk 16]
```diff
@@ -547,21 +607,54 @@ impl SplitAndMergeTag for ID3v2Tag {
...
-			if let Some(total) = split.next() {
-				tag.items
-					.push(TagItem::new(total_key, ItemValue::Text(total.to_string())))
-			}
+			let number_source = split.next()?;
+			let number = parse_number(number_source)?;
+			let total_source = split.next();
+			let total = total_source.and_then(parse_number);
+
+			let is_valid = if let Some(total_source) = total_source {
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:637]: This check doesn't work. You will notice if your add tests for leading/trailing whitespace.

Implicitly assuming that the length of NUMBER_PAIR_SEPARATOR equals 1 byte is brittle.

File: src/id3/v2/tag.rs (hunk 15 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs:612]: Please allow me to confirm the details.

`split_pair()` is called in the `match` statement as the [match guard](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#extra-conditionals-with-match-guards). `if let` guards are experimental now.

https://github.com/rust-lang/rust/issues/51114

Comment (uklotzde) [reply src/id3/v2/tag.rs:612]: Sorry, I wanted to referr to _let-else statements_ introduced with Rust 1.65. These should be convenient here and allow to avoid the final return of a meaningless result.

File: src/id3/v2/tag.rs (hunk 16 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs:637]: Please allow me to confirm the details.

> You will notice if you add tests for leading/trailing whitespace.

What test cases do you think? The followings are passed as a valid content when I added the unit test:

* " 1 "
* "&nbsp;&nbsp;1&nbsp;&nbsp;" (The spaces have 2 characters)
* "1&nbsp;&nbsp;/&nbsp;&nbsp;2" (The spaces have 2 characters)
* "&nbsp;&nbsp;1&nbsp;&nbsp;/&nbsp;&nbsp;2&nbsp;&nbsp;" (The spaces have 2 characters)

Because `content` is checked the following:

* A not trimmed string has a number only.
* One separator character and two not trimmed strings that have a number only.

> Implicitly assuming that the length of NUMBER_PAIR_SEPARATOR equals 1 byte is brittle.

May do not NUMBER_PAIR_SEPARATOR equals 1 bytes? I would be glad if you could provide an example.

Otherwise, do you mean that the constant should be extracted?

File: src/id3/v2/tag.rs [hunk 17]
```diff
@@ -547,21 +607,54 @@ impl SplitAndMergeTag for ID3v2Tag {
...
-			let current = split.next()?.to_string();
-			tag.items
-				.push(TagItem::new(current_key, ItemValue::Text(current)));
 
-			if let Some(total) = split.next() {
-				tag.items
-					.push(TagItem::new(total_key, ItemValue::Text(total.to_string())))
-			}
+			let number_source = split.next()?;
+			let number = parse_number(number_source)?;
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:633]:
```
let Some(number) = parse_number(number_source) else {
    return;
}
```

**Comment by mikanbako** [comment]
@uklotzde Thank you for comments. I replied for them.

File: src/id3/v2/tag.rs (hunk 16 above)
Comment (uklotzde) [reply src/id3/v2/tag.rs:637]: The magic number number 1 appears out of the context of `NUMBER_PAIR_SEPARATOR`. If the constant was of type `str` you could use `NUMBER_PAIR_SEPARATOR.len()`. But using the size of the separator is probably no longer needed after fixing the parsing. It should not be relevant for the decision, only number and total with their parse results need to be considered.

Comment (uklotzde) [reply src/id3/v2/tag.rs:637]: Returning an `Option` from parse_number() is probably no sufficient, because you need to distinguish between _empty_ and _not a number_.

Comment (uklotzde) [reply src/id3/v2/tag.rs:637]: Using the low-level input representation again for making the final decision is conceptually unsound, different levels of abstraction.

Comment (uklotzde) [reply src/id3/v2/tag.rs:637]: Indeed, I overlooked some tests. But I am still unsure that this code covers all cases correctly, especially the expression `content.len() == number_source.len() + 1 + total_source.len()`. I already mentioned the various reasons why it is not suitable here.

File: src/id3/v2/tag.rs [hunk 18]
```diff
@@ -547,21 +607,54 @@ impl SplitAndMergeTag for ID3v2Tag {
...
 			tag: &mut Tag,
-			current_key: ItemKey,
+			number_key: ItemKey,
 			total_key: ItemKey,
 		) -> Option<()> {
+			fn parse_number(source: &str) -> Option<&str> {
+				let number = source.trim();
+
+				if number.is_empty() {
+					return None;
```
Comment (uklotzde) [inline src/id3/v2/tag.rs:617]: Returning an empty `str` (= `Some(number)`) would be an option to distinguish between empty and invalid numbers. There are many possibilities for encoding the 3 variants _empty_, _number_, and _invalid_.

File: src/id3/v2/tag.rs (hunk 16 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs:637]: Thank you for details. I understand that I should improve parsing about `content`. I will consider it.

File: src/id3/v2/tag.rs (hunk 15 above)
Comment (mikanbako) [reply src/id3/v2/tag.rs:612]: Thank you for details. I will consider to use let-else.

**Comment by uklotzde** [comment]
While trying to get rid of the input string comparisons I came up with a simple and effective solution: https://github.com/mikanbako/lofty-rs/pull/1

**Comment by mikanbako** [comment]
@uklotzde Thank you for PR. It is merged.

**Review by uklotzde — APPROVED** [review]
Thank you for this nice cooperation! The code is very solid now after we have considered many possible use cases. The tests will ensure that everything works as expected and are a good starting point if the behavior ever needs to be modified.

@Serial-ATA Please take a final look and then merge. The recent clippy failures seem to be unrelated and should be fixed separately, not by this PR.

**Comment by Serial-ATA** [comment]
Thanks for fixing this! And thanks @uklotzde for the review. This looks good, and has a lot of tests to back it up :smile:.

This makes me realize that the same issue will occur in APE tags as well, which use the same format for track/disk information. I'll have to make an issue for that.
